runtime/stringbuf.c

```diff
@@ -281,6 +281,10 @@ rsRetVal rsCStrAppendStrWithLen(cstr_t *pThis, const uchar* psz, size_t iStrLen)
 	memcpy(pThis->pBuf + pThis->iStrLen, psz, iStrLen);
 	pThis->iStrLen += iStrLen;
 
+	/* reset any cached sz string */
+	free(pThis->pszBuf);
```

Comment:
On Wed, 11 May 2016, Magnus Hyllander wrote:

>> @@ -281,6 +281,10 @@ rsRetVal rsCStrAppendStrWithLen(cstr_t *pThis, const uchar* psz, size_t iStrLen)
>>  	memcpy(pThis->pBuf + pThis->iStrLen, psz, iStrLen);
>>  	pThis->iStrLen += iStrLen;
>>
>> +	/* reset any cached sz string */
>> +	free(pThis->pszBuf);
>
> Please see my comment about `persistStrmState`. The root cause here is that `persistStrmState` calls `rsCStrGetSzStrNoNULL` when only the first line of a multiline message has been read from the stream. This initializes `pszBuf` to the current contents of `pBuf`, which is the first line. When the rest of the lines are appended to `pBuf`, this is not reflected in `pszBuf`. `pszBuf` becomes inconsistent with `pBuf` (shorter), and it results in `MsgSetRawMsg` reading past the end of the allocated memory.
> It's pretty straight-forward to me. I realize this is code from 2005, but it's being used in a new way by imfile that uncovered this bug.

I think that the right answer here isn't to mess with the variables, but rather 
make it so that persistStrmState doesn't get called at each newline, but rather 
it only gets called at the end of a message.

This assumes that we can do this without loosing the data that was just read 
that tells us that we have a new message, not a continuation of the old one.

David Lang